plugins/devlink/fu-devlink-netlink.c

```diff
@@ -0,0 +1,456 @@
+		return MNL_CB_OK;
+
+	tb[type] = attr;
+	return MNL_CB_OK;
+}
+
+static int
+fu_devlink_netlink_genl_mcast_group_attr_cb(const struct nlattr *attr, void *data)
+{
+	const struct nlattr **tb = data;
+	int type = mnl_attr_get_type(attr);
+
+	if (mnl_attr_type_valid(attr, CTRL_ATTR_MCAST_GRP_MAX) < 0)
+		return MNL_CB_OK;
+
+	tb[type] = attr;
+	return MNL_CB_OK;
+}
+
+/* Family resolution callback */
+static int
+fu_devlink_netlink_fu_devlink_netlink_genl_family_get_cb(const struct nlmsghdr *nlh, void *data)
+{
+	FuDevlinkGenSocket *nlg = data;
+	struct genlmsghdr *genl = mnl_nlmsg_get_payload(nlh);
+	struct nlattr *tb[CTRL_ATTR_MAX + 1] = {};
+	struct nlattr *mcgrp;
+
+	mnl_attr_parse(nlh, sizeof(*genl), fu_devlink_netlink_genl_ctrl_attr_cb, tb);
+	if (!tb[CTRL_ATTR_FAMILY_ID])
+		return MNL_CB_ERROR;
+	nlg->family_id = mnl_attr_get_u16(tb[CTRL_ATTR_FAMILY_ID]);
+
+	if (!tb[CTRL_ATTR_MCAST_GROUPS])
+		return MNL_CB_ERROR;
+
+	mnl_attr_for_each_nested(mcgrp, tb[CTRL_ATTR_MCAST_GROUPS])
+	{
+		struct nlattr *tb_grp[CTRL_ATTR_MCAST_GRP_MAX + 1] = {};
+
+		mnl_attr_parse_nested(mcgrp, fu_devlink_netlink_genl_mcast_group_attr_cb, tb_grp);
+
+		if (!tb_grp[CTRL_ATTR_MCAST_GRP_NAME] || !tb_grp[CTRL_ATTR_MCAST_GRP_ID])
+			continue;
+
+		if (!g_strcmp0(mnl_attr_get_str(tb_grp[CTRL_ATTR_MCAST_GRP_NAME]),
+			       DEVLINK_GENL_MCGRP_CONFIG_NAME)) {
+			nlg->config_group_id = mnl_attr_get_u32(tb_grp[CTRL_ATTR_MCAST_GRP_ID]);
+			return MNL_CB_OK;
+		}
+	}
+
+	return MNL_CB_ERROR;
+}
+
+/* Get generic netlink family ID */
+static gboolean
+fu_devlink_netlink_genl_family_get(FuDevlinkGenSocket *nlg,
+				   const gchar *family_name,
+				   GError **error)
+{
+	struct genlmsghdr hdr = {
+	    .cmd = CTRL_CMD_GETFAMILY,
+	    .version = 0x1,
+	};
+	struct nlmsghdr *nlh;
+
+	nlh = fu_devlink_netlink_msg_prepare(nlg->buf, GENL_ID_CTRL, FALSE, &hdr, sizeof(hdr));
+	mnl_attr_put_strz(nlh, CTRL_ATTR_FAMILY_NAME, family_name);
+	return fu_devlink_netlink_msg_send_recv(
+	    nlg,
+	    nlh,
+	    fu_devlink_netlink_fu_devlink_netlink_genl_family_get_cb,
+	    nlg,
+	    error);
+}
+
+/* Open generic netlink socket for devlink family */
+gboolean
+fu_devlink_netlink_gen_socket_open(FuDevlinkGenSocket *nlg, GError **error)
+{
+	gint one = 1;
+	gint rc;
+
+	/* Initialize structure with properly aligned buffer */
+	nlg->buf = g_malloc0(MNL_SOCKET_BUFFER_SIZE);
+
+	/* Open netlink socket */
+	nlg->nl = mnl_socket_open(NETLINK_GENERIC);
+	if (!nlg->nl) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "failed to open netlink socket: %s",
+			    fwupd_strerror(errno));
+		g_free(nlg->buf);
+		return FALSE;
+	}
+
+	rc = mnl_socket_setsockopt(nlg->nl, NETLINK_CAP_ACK, &one, sizeof(one));
+	if (rc < 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "failed to set netlink CAP_ACK: %s",
+			    fwupd_strerror(errno));
+		mnl_socket_close(nlg->nl);
+		g_free(nlg->buf);
```

Comment:
this feels like the kind of thing that needs to be done in the FuDevlinkGenSocket autoptr cleanup